runtime/lua/vim/net.lua

```diff
@@ -0,0 +1,533 @@
+local M = {}
+
+---@class HeaderTable
+---@field private _storage table<string, string[]>
+local HeaderTable = {}
+HeaderTable.__index = HeaderTable
+
+---@private
+function HeaderTable.new(input_table)
+  local instance = setmetatable({ _storage = {} }, HeaderTable)
+  if input_table then
+    instance:_from_table(input_table)
+  end
+  return instance
+end
+
+---@private
+function HeaderTable:_normalize_key(key)
+  return key:lower()
+end
+
+---@private
+function HeaderTable:_from_table(input_table)
+  for key, value in pairs(input_table) do
+    local normalized_key = self:_normalize_key(key)
+    if type(value) == 'string' then
+      self._storage[normalized_key] = { value }
+    elseif type(value) == 'table' then
+      self._storage[normalized_key] = value
+    else
+      vim.notify('Invalid value type for key: ' .. key, vim.log.levels.ERROR)
+    end
+  end
+end
+
+---Set value of header.
+---@param self HeaderTable HeaderTable Instance.
+---@param value string[] | string Header value.
+---@param key string Non case-sensitive header name.
+function HeaderTable:set(key, value)
+  local normalized_key = self:_normalize_key(key)
+  if type(value) == 'string' then
+    self._storage[normalized_key] = { value }
+  elseif type(value) == 'table' then
+    self._storage[normalized_key] = value
+  else
+    vim.notify('Invalid value type for key: ' .. key, vim.log.levels.ERROR)
+  end
+end
+
+---Append value to header.
+---@param self HeaderTable HeaderTable Instance.
+---@param key string Non case-sensitive header name.
+function HeaderTable:append(key, value)
+  local normalized_key = self:_normalize_key(key)
+  if self._storage[normalized_key] then
+    table.insert(self._storage[normalized_key], value)
+  else
+    self._storage[normalized_key] = { value }
+  end
+end
+
+---Get header values.
+---@param self HeaderTable HeaderTable Instance.
+---@param key string Non case-sensitive header name.
+---@return string[] | string | nil
```

Comment:
That's a bad api design tho. Return type should be consistent, it shouldn't return `string` or `string[]`.

For example,

Python:
- `get` returns `string` -https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.get
- `get_all` returns `string[]` - https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.get_all

Javascript:
- `get` returns `string` - https://developer.mozilla.org/en-US/docs/Web/API/Headers/get
- `getSetCookie` returns `string[]` - https://developer.mozilla.org/en-US/docs/Web/API/Headers/getSetCookie

Golang:
- `get` returns `string` - https://pkg.go.dev/net/http#Header.Get
- `values` returns `string[]` - https://pkg.go.dev/net/http#Header.Values 

Rust:
- `get` returns `string` - https://docs.rs/http/latest/http/header/struct.HeaderMap.html#method.get
- `get_all` returns `string` - https://docs.rs/http/latest/http/header/struct.HeaderMap.html#method.get_all

Reply:
Please be more specific in your requests in the future.

Reply:
I'm not telling you to change anything... that's between you and the maintainers. I'm just somebody who's looking forward to this feature and so giving feedback from the user's perspective.

And about being specific,

> I'd expect `:get()` to return the `string` like how it's actually transported in the http header.

HTTP header values are just plain strings. Even if a header has multiple values they're all joined together in a single string when being transported over http. Not sure if I could be more specific.